components/scream/src/physics/shoc/shoc_inputs_initializer.cpp

```diff
@@ -0,0 +1,354 @@
+#include "physics/shoc/shoc_inputs_initializer.hpp"
+#include "physics/shoc/shoc_main_impl.hpp"
+#include "ekat/util/ekat_file_utils.hpp"
+#include "physics/share/physics_functions.hpp"
+
+#include <array>
+#include <fstream>
+
+namespace scream
+{
+
+void SHOCInputsInitializer::add_field (const field_type &f)
+{
+  const auto& id = f.get_header().get_identifier();
+  
+  m_fields.emplace(id.name(),f);
+  m_fields_id.insert(id);
+}
+
+void SHOCInputsInitializer::
+add_field (const field_type &f, const field_type& f_ref,
+           const remapper_ptr_type& remapper)
+{
+  if (m_remapper) {
+    // Sanity check
+    EKAT_REQUIRE_MSG (m_remapper->get_src_grid()->name()==remapper->get_src_grid()->name(),
+      "Error! A remapper was already set in SHOCInputsInitializer, but its src grid differs from"
+      "       the grid of the input remapper of this call.\n");
+  } else {
+    m_remapper = remapper;
+    m_remapper->registration_begins();
+  }
+
+  const auto& id = f.get_header().get_identifier();
+  const auto& id_ref = f_ref.get_header().get_identifier();
+
+  // To the AD, we only expose the fact that we init f_ref...
+  m_fields_id.insert(id_ref);
+
+  // ...but SHOC only knows how to init f...
+  m_fields.emplace(id.name(),f);
+
+  // ...hence, we remap to f_ref.
```

Comment:
These are lovely comments. Thanks!

Reply:
I think we must credit @AaronDonahue for using it in p3_inputs_initializer...